src/srkSmtlib2.ml

```diff
@@ -0,0 +1,624 @@
+          let sym = of_symbol srk_sym in
+          match typ_symbol srk sym with
+          | `TyBool | `TyInt | `TyReal -> invalid_arg "Unexpected function application: Non-functional symbol"
+          | `TyFun (args_typ, _) ->
+            if (List.length terms) != (List.length args_typ) then invalid_arg "Unexpected function application: arity mistmatch" else
+            let args =
+              List.map2 (fun term typ ->
+                match typ with
+                | `TyBool -> (formula_of term :> 'a gexpr)
+                | _ -> (term_of term :> 'a gexpr)
+              ) terms args_typ
+            in
+            (mk_app srk sym args :> 'a gexpr)
+        end
+      end
+    | App ((("divisible", [Num n]), None), terms) ->
+      begin match terms with
+      | [t] ->
+        let t = term_of t in
+        let sym = mk_symbol srk `TyInt in
+        let n = mk_zz srk n in
+        (mk_exists_const srk sym (mk_eq srk t (mk_mul srk [n; mk_const srk sym])) :> 'a gexpr)
+      | _ -> assert false
+      end
+  (*  | Let (bindings, t) -> invalid_arg "Unsupported"
+    | Forall (vars, t) -> invalid_arg "Unsupported"
+    | Exists (vars, t) -> invalid_arg "Unsupported" *)
+    | App _ -> invalid_arg "Unknown function name"
+    | _ -> invalid_arg "Match and attribute statments are not supported"
+  and term_of_smt of_symbol t =
+    match Syntax.Expr.refine srk (expr_of_smt of_symbol t) with
+    | `Term t -> t
+    | _ -> invalid_arg "Term expected"
+  and formula_of_smt of_symbol t =
+    match Syntax.Expr.refine srk (expr_of_smt of_symbol t) with
+    | `Formula phi -> phi
+    | _ -> invalid_arg "Formula expected"
+
+  type symbol_pool = {
+    new_symbol : Syntax.typ -> Syntax.symbol;
+    reset : unit -> unit
+  }
+
+  (* pool of symbols for each base type: avoids possible memory leak *)
+  let formals_pool : symbol_pool =
+    let module DA = BatDynArray in
+    let int_symbols = DA.create () in
+    let int_id = ref (-1) in
+    let real_symbols = DA.create () in
+    let real_id = ref (-1) in
+    let bool_symbols = DA.create () in
+    let bool_id = ref (-1) in
+    let new_symbol typ =
+      let open Syntax in
+      match typ with
+      | `TyInt -> incr int_id;
+        if (!int_id) < (DA.length int_symbols) then DA.get int_symbols (!int_id) else
+        begin let sym = mk_symbol srk typ in DA.add int_symbols sym; sym end
+      | `TyReal -> incr real_id;
+        if (!real_id) < (DA.length real_symbols) then DA.get real_symbols (!int_id) else
+        begin let sym = mk_symbol srk typ in DA.add real_symbols sym; sym end
+      | `TyBool -> incr bool_id;
+        if (!bool_id) < (DA.length bool_symbols) then DA.get bool_symbols (!int_id) else
+        begin let sym = mk_symbol srk typ in DA.add bool_symbols sym; sym end
+      | _ -> assert false
+    in
+    let reset _ =
+      int_id := -1;
+      real_id := -1;
+      bool_id := -1
+    in
+    { new_symbol; reset }
+
+  let rec simplify_const_expr expr =
```

Comment:
This looks like it belongs in srkSimplify.ml.  Is there functionality in this simplfier that isn't present in the srkSimplify simplification routines / SimplifyingContext in syntax.ml?

Reply:
You're right. This code does constant folding, which is covered by SimplifyingContext. Although, on careful examination, it seems the simplifying code in SimplifyingContext and this code differ on Mod.

Specifically, I think there's a typo in the SimplifyingContext code. It checks if the denominator is zero twice. I think the second check should actually be if the denominator is one and return zero. I can post a fix for that, but I have a question, do we support non-integral semantics for mod? e.g. can you do 1.2 mod 1 = 0.2?